File: src/utils/google-slides-url.ts
export type ParsedSlidesUrl = {
  /** Only present for "source" edit URLs - this is what the Slides API needs. */
  presentationId: string | null;
  /** True for /presentation/d/e/<publishedId>/... URLs, whose ID cannot be used with the API. */
  isPublished: boolean;
};

const SOURCE_ID_PATTERN = /\/presentation\/d\/([a-zA-Z0-9_-]+)/;
const PUBLISHED_ID_PATTERN = /\/presentation\/d\/e\/([a-zA-Z0-9_-]+)/;

export function isValidGoogleSlidesUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    return parsed.hostname === "docs.google.com" && parsed.pathname.includes("/presentation/");
  } catch {
    return false;
  }
}

/**
 * Source ("edit") URLs and published ("pub"/"embed") URLs encode different, mutually
 * incompatible identifiers - a published ID cannot be used with the Google Slides API.
 * This is why LessonConfig stores slidesSourceUrl and slidesEmbedUrl separately.
 */
export function parseGoogleSlidesUrl(url: string): ParsedSlidesUrl {
  const publishedMatch = url.match(PUBLISHED_ID_PATTERN);
  if (publishedMatch) {
    return { presentationId: null, isPublished: true };
  }
  const sourceMatch = url.match(SOURCE_ID_PATTERN);
  if (sourceMatch) {
    return { presentationId: sourceMatch[1], isPublished: false };
  }
  return { presentationId: null, isPublished: false };
}

export function buildEmbedUrlFromPresentationId(presentationId: string): string {
  return `https://docs.google.com/presentation/d/${presentationId}/embed?start=false&loop=false&delayms=60000`;
}

File: backend/src/SupportRoom.Providers.Slides/GoogleSlidesUrl.cs
using System.Text.RegularExpressions;

namespace SupportRoom.Providers.Slides;

public readonly record struct ParsedSlidesUrl(string? PresentationId, bool IsPublished);

/// <summary>Mirrors src/utils/google-slides-url.ts.</summary>
public static partial class GoogleSlidesUrl
{
    [GeneratedRegex(@"/presentation/d/([a-zA-Z0-9_-]+)")]
    private static partial Regex SourceIdPattern();

    [GeneratedRegex(@"/presentation/d/e/([a-zA-Z0-9_-]+)")]
    private static partial Regex PublishedIdPattern();

    public static bool IsValid(string url)
    {
        return Uri.TryCreate(url, UriKind.Absolute, out var parsed)
            && parsed.Host.Equals("docs.google.com", StringComparison.OrdinalIgnoreCase)
            && parsed.AbsolutePath.Contains("/presentation/");
    }

    /// <summary>
    /// Source ("edit") URLs and published ("pub"/"embed") URLs encode different, mutually
    /// incompatible identifiers - a published ID cannot be used with the Google Slides API.
    /// </summary>
    public static ParsedSlidesUrl Parse(string url)
    {
        if (PublishedIdPattern().IsMatch(url))
        {
            return new ParsedSlidesUrl(null, true);
        }

        var sourceMatch = SourceIdPattern().Match(url);
        return sourceMatch.Success
            ? new ParsedSlidesUrl(sourceMatch.Groups[1].Value, false)
            : new ParsedSlidesUrl(null, false);
    }

    public static string BuildEmbedUrl(string presentationId)
        => $"https://docs.google.com/presentation/d/{presentationId}/embed?start=false&loop=false&delayms=60000";
}
